README.md
# DeveloPic
존립을 위협받고 있는 이름이지만 일단 어쩔 수 없으니 유지

## 서버 실행 방법
`CHATGPT_API_KEY` 환경변수를 OpenAI API Key로 설정. Key는 프로젝트 리더에게서 발급받거나 개인 계정 사용.
```bash
export CHATGPT_API_KEY=...
```

다음 명령어로 서버를 실행 (Go 1.20.2 이상의 버전이 설치되어 있어야 함)
```bash
cd backend/gateway/cmd && go run main.go
```

`localhost:3000` 으로 접속하면 API 이용 가능
```bash
curl --request GET -sL \
     --url 'http://localhost:3000/problems/234'
```


front/history.md
## 2023

### 4/1

-   개발환경 세팅 완료.
    -   mui5, iconify, lint ...
    -   v0.0.0:1

### 4/2

-   api 사용
-   flowbite 적용
-   v0.0.0:2

### 4/3

-   api key 적용
-   v0.0.0:3

### 4/4

-   problem 적용
-   v0.0.0:4
-   언어 선택하여 problem 사용
-   v0.0.0:5

### 4/7

-   codeMirror 적용
-   v0.0.0:6
-   codeMirror 테마 적용
-   v0.0.0:7

### 4/10

-   design pattern 적용
-   navBar 적용
-   v0.0.0:8

### 4/11

-   resizerble 제작 중
-   v0.0.0:8
-   resizerble 프로토타입 제작 완료
-   v0.0.0:9

### 4/12

-   api 연동 후 base64 데이터 받음.
-   base64 데이터 받고 md로 출력. - 조정 중
-   v0.0.0:10

### 4/14

-   reactMarkDown 적용 완.
-   v0.0.0:11

### 5/6

-   router 적용 중 개같은 eslint
-   v0.0.0:12

### 5/10

-   페이지 3개로 변경
-   v0.0.0:13

### 5/12

-   stepper 제작
-   v0.0.0:14
-   router 이동 시 모션 적용.
-   v0.0.0:15

### 5/14

-   api 사용 준비
-   v0.0.0:15

### 5/15

-   recoil 적용
-   v0.0.0:16
-   Cumtom hook으로 변경해야함.
-   v0.0.0:17

### 5/16

-   coustom hook으로 Stepper 리팩토링
-   v0.0.0:18

### 5/17

-   problem, stepper 리팩토링 완료
-   v0.0.0:19

### 5/18

-   msw 적용
-   v0.0.0:20
-   suspence 적용
-   resizable hook 제작
-   v0.0.0:21
-   msw가 똑같이 작동하도록 개선 중.
-   recoil적용을 위해 Problem 페이지 컴포넌트화 필요
-   v0.0.0:22

### 5/19

-   suspence가 api를 인식하도록 개선 중.
-   v0.0.0:23

### 5/20

-   개같은 Suspence 포기
-   v0.0.0:24

### 5/21

-   lint error 개선, suspence 이해
-   v0.0.0:25

### 5/24

-   suspece 잘 작동하는 것 확인
-   v0.0.0:26
-   recoil 작성 중
-   v0.0.0:27

### 5/26

-   서스펜스 제거하고 stepper화면 따로 라우팅하여 구현
-   v0.0.0:28
-   디자인시스템 적용
-   메인화면 제작 중
-   v0.0.0:29

### 5/27

-   select box 제작
-   v0.0.0:30
-   select box에 대응하는 버튼 제작
-   v0.0.0:31

### 5/30

-   codeEditor, markdown 개선
-   v0.0.0:32

### 6/6

-   typeError 전부 제거!
-   v0.0.0:33
-   select 클릭 안되는 부분 개선
-   color 디자인 시스템 적용.
-   v0.0.0:34
-   button 디자인 시스템 적용
-   v0.0.0:35
-   stepper 버그 제거
-   api cancel 추가
-   v0.0.0:36

### 6/9

-   iconfont component 제작
-   profile icon 추가
-   v0.0.0:37
-   problem nav top 추가
-   button function 제작 중.
-   v0.0.0:38

### 6/10

-   openapi 변경 사항 적용
-   problem.hook 리팩토링
-   v0.0.0:39

### 6/11

-   logo, favicon,profile 변경
-   iconify 제거
-   v0.0.0:40
-   styled component 이용하여 button 제작
-   v0.0.0:41
-   problem 페이지
    nav바 디자인 적용
-   v0.0.0:42

### 6/13

-   codeEditor overflow 버그 개선
-   resizable 제거
-   v0.0.0:43
-   problem.hook useEffect 문제 해결하면서 생긴 사이드 이펙트 제거
-   v0.0.0:44

### 6/14

-   problem localstorage 적용
-   v0.0.0:45
-   reset 버튼 기능 제작
-   v0.0.0:46
-   problem nav의 언어,난이도 로컬스토리지 적용
-   v0.0.0:47
-   home 접속 시 옵션 초기화
-   v0.0.0:48

### 6/17

-   api 변경사항에 따라 msw 수정 & 추가
-   v0.0.0:49
-   profile, result router 적용
-   v0.0.0:50
-   result 페이지 완성
-   v0.0.0:51
-   profile 페이지 완성
-   v0.0.0:52
-   profile api 적용
-   v0.0.0:53

### 6/18

-   api header 전부 profile 헤더로 적용
-   v0.0.0:54
-   submit 버튼 기능 완료
-   v0.0.0:55
-   간이 로딩 적용
-   v0.0.0:56
-   more 버튼 비활성화, hook 에러 수정 -> 최상위에 놓지 않아서 생김.
-   v0.0.0:57
-   profile 이미지 svg로 변경
-   v0.0.0:58
-   problem 페이지에 접속할 때 problem 값이 없으면 home으로 리다이렉트
-   v0.0.0:59

### 6/19

-   font 적용
-   문제 타이틀, 설명 ui 개선
-   v0.0.0:60

### 6/23

- profile 스코어없으면 없다고 표시
- v0.0.0:61
- e2e 테스트 home 제작
- v0.0.0:62
- 코드 창 짤리는 버그 제거
- v0.0.0:63
- 언어별로 code다르게 받도록 구현
- localstorage 관리가 복잡해 유저플로우 구현 후 재구성 해야할 듯.
- v0.0.0:64
- codeEditor go언어 적용
- v0.0.0:65

backend/main.go
package main

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"github.com/labstack/echo/v4"
	"github.com/labstack/echo/v4/middleware"
	"github.com/sashabaranov/go-openai"
	"log"
	"os"
)

type Body struct {
	Statement string   `json:"statement"`
	Points    []string `json:"points"`
	Code      string   `json:"code"`
}

type ProblemRequestBody struct {
	Difficulty int32  `json:"difficulty"`
	Language   string `json:"language"`
}

type ProblemResponseBody struct {
	Statement   string    `json:"statement"`
	Description string    `json:"description"`
	Examples    []Example `json:"examples"`
}

type Example struct {
	Input  any `json:"input"`
	Output any `json:"output"`
}

func main() {
	apiKey := os.Getenv("CHATGPT_API_KEY")
	cli := openai.NewClient(apiKey)
	app := echo.New()
	app.Use(middleware.CORS())

	gptClient := ChatGPTClient{cli: cli}

	app.POST("/problem", func(c echo.Context) error {
		var body ProblemRequestBody
		if err := c.Bind(&body); err != nil {
			return err
		}

		if body.Difficulty == 0 {
			body.Difficulty = 50
		}

		if body.Language == "" {
			body.Language = "Go"
		}

		sample := ProblemResponseBody{
			Statement:   "Powers of Two",
			Description: "Have the function `PowersofTwo(num)` take the num parameter being passed which will be an integer and return the string true if it's a power of two. If it's not return the string false. For example if the input is 16 then your program should return the string true but if the input is 22 then the output should be the string false.",
			Examples: []Example{
				{
					Input:  "4",
					Output: "true",
				},
				{
					Input:  "124",
					Output: "false",
				},
			},
		}
		sampleJSON, _ := json.MarshalIndent(sample, "", "  ")

		messages := []openai.ChatCompletionMessage{
			{
				Role:    openai.ChatMessageRoleUser,
				Content: fmt.Sprintf("Make a real world scenario briefly for a developer. The complexity should be %d out of 100. Complexity is a measure of how complex a problem situation is. The higher it is, the more complicated the situation. Exclude the AI technology.", body.Difficulty),
			},
			{
				Role:    openai.ChatMessageRoleUser,
				Content: fmt.Sprintf("Using the previous scenario, select one functionality and make only one coding test to implement it in %s. It should evaluate the technical skills, problem solving and analytical skills of the developer. The difficulty is %d out of 100. You should give at least two examples.", body.Language, body.Difficulty),
			},
			{
				Role:    openai.ChatMessageRoleUser,
				Content: "Don't use the markdown syntax. You can just use the JSON syntax. Don't include your words.",
			},
			{
				Role:    openai.ChatMessageRoleUser,
				Content: "Convert the result to JSON like this:",
			},
			{
				Role:    openai.ChatMessageRoleUser,
				Content: string(sampleJSON),
			},
		}

		msg, err := gptClient.CompleteChatWithContext(c.Request().Context(), messages)
		if err != nil {
			return err
		}

		//messages = []openai.ChatCompletionMessage{
		//
		//}
		//
		//msg, err := gptClient.CompleteChatWithContext(c.Request().Context(), messages)
		//if err != nil {
		//	return err
		//}

		log.Println(msg)

		var response ProblemResponseBody
		if err := json.Unmarshal([]byte(msg), &response); err != nil {
			log.Println(err)
			return err
		}

		return c.JSON(201, response)
	})

	app.POST("/analyze", func(c echo.Context) error {
		var body Body
		if err := c.Bind(&body); err != nil {
			return err
		}

		decoded, err := base64.StdEncoding.DecodeString(body.Code)
		if err != nil {
			return err
		}

		prompt := `
Please evaluate the characteristics of the programmer based on the following code and give the result in a JSON format.
You must consider the following statement and points when evaluating the code.
`

		// Analyze the code
		response, err := cli.CreateChatCompletion(c.Request().Context(), openai.ChatCompletionRequest{
			Model:       openai.GPT3Dot5Turbo0301,
			Temperature: 1,
			Messages: []openai.ChatCompletionMessage{
				{
					Role:    openai.ChatMessageRoleUser,
					Content: prompt,
				},
				{
					Role:    openai.ChatMessageRoleUser,
					Content: string(decoded),
				},
				{
					Role:    openai.ChatMessageRoleUser,
					Content: fmt.Sprintf("The statement is %s and the points are %v.", body.Statement, body.Points),
				},
				{
					Role:    openai.ChatMessageRoleUser,
					Content: "Convert the result to JSON format has four keys",
				},
				{
					Role:    openai.ChatMessageRoleUser,
					Content: "The first key is keyword which is array of string type that has 5 short keywords which describe the developer's characteristics well.",
				},
				{
					Role:    openai.ChatMessageRoleUser,
					Content: "The second key is score which is integer type that has the score of the code out of 100.0",
				},
				{
					Role:    openai.ChatMessageRoleUser,
					Content: "The third key is strengths which is array of string type that has 5 elements which have the strengths of the code",
				},
				{
					Role:    openai.ChatMessageRoleUser,
					Content: "The fourth key is weaknesses which is array of string type that has 5 elements which have the weaknesses of the code",
				},
			},
		})
		if err != nil {
			return err
		}

		log.Println(response.Choices[0].Message.Content)

		var responseMap map[string]interface{}
		if err := json.Unmarshal([]byte(response.Choices[0].Message.Content), &responseMap); err != nil {
			log.Println(err)
			return err
		}

		return c.JSON(200, responseMap)
	})

	if err := app.Start(":8080"); err != nil {
		panic(err)
	}
}

type ChatGPTClient struct {
	cli      *openai.Client
	messages []openai.ChatCompletionMessage
}

func (c *ChatGPTClient) CompleteChatWithContext(ctx context.Context, messages []openai.ChatCompletionMessage) (string, error) {
	c.messages = append(c.messages, messages...)

	response, err := c.cli.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
		Model:       openai.GPT3Dot5Turbo0301,
		Temperature: 1,
		Messages:    c.messages,
	})
	if err != nil {
		return "", err
	}

	c.messages = append(c.messages, response.Choices[0].Message)

	return response.Choices[0].Message.Content, nil
}
